More work on the docs (#41)
nextflow-io/nf-validation: docs/samplesheets/immutable_map.md

---
description: Functions and methods for the ImmutableMap class
---

# Immutable meta maps

## Introduction and usage

> A nasty trap when using meta maps in Nextflow pipelines is to forget that they are mutable.
> That is; if you edit a value then it will persist for other processes.

Reviewer: ```suggestion
A common pitfall when utilizing meta maps in Nextflow pipelines is overlooking their mutable nature. Meta maps in Nextflow are mutable, meaning that if you change a value inside the map, this change will persist across different processes and channels in your pipeline. This can lead to unexpected behavior if not handled correctly.
```

This `ImmutableMap` object is returned by default, but if you prefer you can return a normal `LinkedHashMap` by setting the `fromSamplesheet()` option `immutable_meta` to `false`.

```groovy
Channel.fromSamplesheet(
  'input',
  immutable_meta: false
)
```

The `immutable_meta` setting can also be set by pipeline users at run time, with the pipeline parameter `params.validationImmutableMeta`.
This allows end-users to override the setting in case of problems launching the pipeline.

```bash
nextflow run <pipeline> --validationImmutableMeta
```

!!! info

    `nextflow.validation.ImmutableMap` is an extension of the `java.util.LinkedHashMap` class. It inherits all the methods from it's parent class, but changes and adds some methods.

## New methods

The following custom new methods have been added to the `ImmutableMap` class:

1. `dropKeys(<Collection>)`: Returns a new `ImmutableMap` without the specified keys
2. `dropKey(<String>)`: Returns a new `ImmutableMap` without the specified key

## Methods that return a copy of the map

These methods return a _new_ instance of `ImmutableMap`.
This copy is safe to work with as it does not affect the original meta map.

> - `collectEntries()`

Reviewer: a one line description about these methods as I feel they are quite new to beginners ?

Author: Maybe, or I wondered about cutting it back. These are native groovy methods, we're just saying that we tweak how they work. 99% of people using this won't need to care about this.

Reviewer 2: Maybe we could make this some kind of drop down menu? So only users that really want to know it have to look at it?

Reviewer: i would like them to stay, +1 for a drop-down if possible.

Author: A drop down isn't easy (website is built with [MkDocs Material](https://squidfunk.github.io/mkdocs-material/reference/). I used a bit of custom CSS to compress them so that they take up less room. They look a little neater now.

> - `collectEntries()`
> - `drop()`
> - `dropWhile()`
> - `each()`
> - `eachWithIndex()`
> - `findAll()`
> - `groupBy()`
> - `groupEntriesBy()`
> - `intersect()`
> - `leftShift()`
> - `minus()`
> - `plus()`
> - `reverseEach()`
> - `sort()`
> - `subMap()`
> - `take()`
> - `takeWhile()`
> - `toSorted()`
> - `withDefault()`

Reviewer 3: Suggesting a dropdown as mentioned [here](https://github.com/nextflow-io/nf-validation/pull/41/files#r1214038200)
```suggestion
<details>
<summary>See a list of the methods</summary>
- `collectEntries()`
- `drop()`
- `dropWhile()`
- `each()`
- `eachWithIndex()`
- `findAll()`
- `groupBy()`
- `groupEntriesBy()`
- `intersect()`
- `leftShift()`
- `minus()`
- `plus()`
- `reverseEach()`
- `sort()`
- `subMap()`
- `take()`
- `takeWhile()`
- `toSorted()`
- `withDefault()`
</details>
```

Author: ahh, _that_ kind of drop down. I was imagining some kind of `<select>` drop-down, or search interface.

I squished these up with some CSS to make them more compact, I think that works fine without needing to hide them under a `<details>`, but thanks anyway 👍🏻

## Methods that return an error

The following methods _change_ the contents of the map and have been disabled.
If you use them in a pipeline it will exit with an error.

> - `put()`
> - `putAll()`
> - `remove()`
> - `clear()`

Reviewer 3: ```suggestion
<details>
<summary>See a list of the methods</summary>
- `put()`
- `putAll()`
- `remove()`
- `clear()`
</details>
```
